### PR TITLE
Introduce DockerContainer#Builder

### PR DESCRIPTION
**Pull request by kokosing**
Introduce DockerContainer#Builder

### CODE REVIEW

File: presto-testing-docker/src/main/java/io/prestosql/testing/docker/DockerContainer.java [hunk 1]
```diff
@@ -63,20 +63,20 @@
...
     private Map<Integer, Integer> hostPorts;
 
-    public DockerContainer(String image, List<Integer> ports, Map<String, String> environment, CheckedConsumer<HostPortProvider> healthCheck)
+    public DockerContainer(String image, List<Integer> ports, Map<String, String> environment, List<String> capabilities, CheckedConsumer<HostPortProvider> healthCheck)
     {
         this.image = requireNonNull(image, "image is null");
         this.environment = ImmutableMap.copyOf(requireNonNull(environment, "environment is null"));
         try {
-            startContainer(ports, healthCheck);
+            startContainer(ports, healthCheck, capabilities);
```
Comment (findepi) [inline presto-testing-docker/src/main/java/io/prestosql/testing/docker/DockerContainer.java]: Assign it to a field instead of passing around.
Also validate it's not null and doesn't contain nulls (`ImmutableList.copyOf`), instead of relying on docker lib's `.capAdd` doing the validation.

File: presto-mysql/src/test/java/io/prestosql/plugin/mysql/DockerizedMySqlServer.java [hunk 2]
```diff
@@ -48,6 +48,7 @@ public DockerizedMySqlServer()
                         "MYSQL_USER", MYSQL_USER,
                         "MYSQL_PASSWORD", MYSQL_PASSWORD,
                         "MYSQL_DATABASE", "tpch"),
+                ImmutableList.of(),
```
Comment (findepi) [inline presto-mysql/src/test/java/io/prestosql/plugin/mysql/DockerizedMySqlServer.java]: nit:
change cmt msg `Allow to set capabilities for DockerContainer` to `Allow setting capabilities in DockerContainer`

File: presto-testing-docker/src/main/java/io/prestosql/testing/docker/DockerContainer.java [hunk 3]
```diff
@@ -132,10 +135,10 @@ private void createContainer(List<Integer> ports)
         Set<String> exposedPorts = ports.stream()
                 .map(Object::toString)
                 .collect(toImmutableSet());
-
```
Comment (findepi) [inline presto-testing-docker/src/main/java/io/prestosql/testing/docker/DockerContainer.java]: unrelated

File: presto-testing-docker/src/main/java/io/prestosql/testing/docker/DockerContainer.java [hunk 4]
```diff
@@ -56,27 +55,30 @@
...
+            startContainer(image, ports, environment, capabilities, healthCheck);
         }
         catch (Exception e) {
             closeAllSuppress(e, this);
             throw new RuntimeException(e);
         }
     }
 
-    private void startContainer(List<Integer> ports, CheckedConsumer<HostPortProvider> healthCheck, List<String> capabilities)
+    private void startContainer(String image, List<Integer> ports, Map<String, String> environment, List<String> capabilities, CheckedConsumer<HostPortProvider> healthCheck)
```
Comment (findepi) [inline presto-testing-docker/src/main/java/io/prestosql/testing/docker/DockerContainer.java]: I'd prefer to go into opposite direction. `DockerContainer` is stateful anyway and it doesn't help it if we pass configuration explicitly to methods.

File: presto-testing-docker/src/main/java/io/prestosql/testing/docker/DockerContainer.java [hunk 5]
```diff
@@ -253,4 +255,54 @@ public void close()
...
+        public Builder(String image)
+        {
+            this.image = requireNonNull(image, "image is null");
+        }
+
+        public Builder setPorts(int... ports)
+        {
+            this.ports = IntStream.of(ports)
+                    .boxed()
+                    .collect(toImmutableList());
```
Comment (findepi) [inline presto-testing-docker/src/main/java/io/prestosql/testing/docker/DockerContainer.java]:
```
this.ports = ImmutableList.copyOf(com.google.common.primitives.Ints.asList(ports));
```

(+ `rnn` if you want to keep `rnn` in the other setters)

File: presto-testing-docker/src/main/java/io/prestosql/testing/docker/DockerContainer.java [hunk 6]
```diff
@@ -253,4 +255,54 @@ public void close()
...
+
+        public Builder setCapabilities(String... capabilities)
+        {
+            this.capabilities = ImmutableList.copyOf(requireNonNull(capabilities, "capabilities is null"));
+            return this;
+        }
+
+        public Builder setHealthCheck(CheckedConsumer<HostPortProvider> healthCheck)
+        {
+            this.healthCheck = healthCheck;
```
Comment (findepi) [inline presto-testing-docker/src/main/java/io/prestosql/testing/docker/DockerContainer.java]: `rnn` like in the others?

File: presto-mysql/src/test/java/io/prestosql/plugin/mysql/DockerizedMySqlServer.java [hunk 7]
```diff
@@ -40,16 +39,15 @@
 
     public DockerizedMySqlServer()
     {
-        this.dockerContainer = new DockerContainer(
-                "mysql:8.0.15",
-                ImmutableList.of(MYSQL_PORT),
-                ImmutableMap.of(
+        this.dockerContainer = DockerContainer.builder("mysql:8.0.15")
```
Comment (findepi) [inline presto-mysql/src/test/java/io/prestosql/plugin/mysql/DockerizedMySqlServer.java]: It would look better if `image` was provider via regular setter. (I'm aware it's a required param.)

Comment (electrum) [reply presto-mysql/src/test/java/io/prestosql/plugin/mysql/DockerizedMySqlServer.java]: Maybe we could rename this method?
```
this.dockerContainer = DockerContainer.image("mysql:8.0.15")
        .setPorts(MYSQL_PORT)
        .setHealthCheck(DockerizedMySqlServer::healthCheck)
        .build();

Comment (findepi) [reply presto-mysql/src/test/java/io/prestosql/plugin/mysql/DockerizedMySqlServer.java]: Might do. `forImage`?

File: presto-testing-docker/src/main/java/io/prestosql/testing/docker/DockerContainer.java [hunk 8]
```diff
@@ -248,4 +255,54 @@ public void close()
...
+            return this;
+        }
+
+        public DockerContainer build()
```
Comment (findepi) [inline presto-testing-docker/src/main/java/io/prestosql/testing/docker/DockerContainer.java]: `start()`

File: presto-mysql/src/test/java/io/prestosql/plugin/mysql/DockerizedMySqlServer.java (hunk 7 above)
Comment (kokosing) [reply presto-mysql/src/test/java/io/prestosql/plugin/mysql/DockerizedMySqlServer.java]: Thanks, it looks much better now.

File: presto-testing-docker/src/main/java/io/prestosql/testing/docker/DockerContainer.java [hunk 9]
```diff
@@ -69,16 +70,17 @@ public DockerContainer(String image, List<Integer> ports, Map<String, String> en
         this.image = requireNonNull(image, "image is null");
         this.environment = ImmutableMap.copyOf(requireNonNull(environment, "environment is null"));
         this.capabilities = ImmutableList.copyOf(requireNonNull(capabilities, "capabilities is null"));
+        this.healthCheck = requireNonNull(healthCheck, "healthCheck is null");
         try {
-            startContainer(ports, healthCheck);
+            startContainer(ports);
```
Comment (findepi) [inline presto-testing-docker/src/main/java/io/prestosql/testing/docker/DockerContainer.java]: > `Store healthCheck as field`

nit: why not store `ports` as a field too?

**Comment by findepi** [comment]
> `Introduce DockerContainer#Builder`

nit: consider changing title to `Introduce builder for DockerContainer`

File: presto-testing-docker/src/main/java/io/prestosql/testing/docker/DockerContainer.java (hunk 9 above)
Comment (kokosing) [reply presto-testing-docker/src/main/java/io/prestosql/testing/docker/DockerContainer.java]: because it is already stored as `hostPorts` (map)
